odl/tomo/operators/ray_trafo.py

```diff
@@ -273,299 +158,217 @@ def __init__(self, reco_space, geometry, variant, **kwargs):
+                    raise ValueError(
+                        '{!r} back-end not available'.format(impl))
+
+                impl_type = RAY_TRAFO_IMPLS[impl.lower()]
+            elif isinstance(impl, type) or isinstance(impl, object):
+                # User gave the type and leaves instantiation to us
+                forward = getattr(impl, "call_forward", None)
+                backward = getattr(impl, "call_backward", None)
+
+                if not callable(forward) and not callable(backward):
+                    raise TypeError('Type {!r} must be have a `call_forward` '
+                                    'or `call_backward`.'.format(impl))
+
+                if isinstance(impl, type):
+                    impl_type = impl
                 else:
-                    astra_wrapper = self._astra_wrapper
-
-                return astra_wrapper.call_forward(x_real, out_real, **kwargs)
+                    # User gave an object for `impl`, meaning to set the
+                    # backend cache to an already initiated object
+                    impl_type = type(impl)
+                    impl_instance = impl
             else:
-                # Should never happen
-                raise RuntimeError('bad `impl` {!r}'.format(self.impl))
+                raise TypeError(
+                    '`impl` {!r} should be a `str`, or an object or type '
+                    'having a `call_forward()` and/or `call_backward()`. '
+                    ''.format(type(impl)))
 
-        elif self.impl == 'skimage':
-            return skimage_radon_forward_projector(
-                x_real, self.geometry, self.range.real_space, out_real,
-                **kwargs)
-        else:
-            # Should never happen
-            raise RuntimeError('bad `impl` {!r}'.format(self.impl))
+        return impl_type, impl_instance
 
     @property
-    def adjoint(self):
-        """Adjoint of this operator.
-
-        Returns
-        -------
-        adjoint : `RayBackProjection`
-        """
-        if self._adjoint is not None:
-            return self._adjoint
-
-        kwargs = self._extra_kwargs.copy()
-        kwargs['domain'] = self.range
-        self._adjoint = RayBackProjection(self.domain, self.geometry,
-                                          impl=self.impl,
-                                          use_cache=self.use_cache,
-                                          **kwargs)
-        return self._adjoint
-
+    def impl(self):
+        """Implementation name string.
 
-class RayBackProjection(RayTransformBase):
+        If a custom ``impl`` was provided this method returns a ``str``
+        of the type."""
 
-    """Adjoint of the discrete Ray transform between L^p spaces."""
+        return self.__impl
 
-    def __init__(self, range, geometry, **kwargs):
-        """Initialize a new instance.
+    def create_impl(self, use_cache=True):
+        """Fetches or instantiates implementation backend for evaluation.
```

Comment:
Those two cases are perhaps best summarized by the name `get`.

Reply:
For me that reminds me of the dumb getters/setters of many other languages, but that might not be a valid reason. Changed it.